### PR TITLE
executor: fix the potential goroutine leak in cluster log retriever

### LINKED ISSUE #14747: clusterLogRetriever fails w/ leak test
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
[2020-02-12T08:16:31.405Z] --- FAIL: TestT (13.57s)
[2020-02-12T08:16:31.405Z]     leaktest.go:143: Test TestT check-count 50 appears to have leaked: github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving.func1.1()
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:534 +0x424
[2020-02-12T08:16:31.405Z]         github.com/pingcap/tidb/util.WithRecovery(0xc019d6f728, 0x0)
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/util/misc.go:86 +0x57
[2020-02-12T08:16:31.405Z]         github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving.func1(0xc01bae83c0, 0xc01bae83b0, 0xc01bc00d20, 0xc01afe60c0, 0xc00516d3fb, 0x2, 0xc00516d3fe, 0xf, 0xc00516d40e, 0xf)
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:506 +0xeb
[2020-02-12T08:16:31.405Z]         created by github.com/pingcap/tidb/executor.(*clusterLogRetriever).startRetrieving
[2020-02-12T08:16:31.405Z]         	/home/jenkins/agent/workspace/tidb_ghpr_unit_test/go/src/github.com/pingcap/tidb/executor/cluster_reader.go:505 +0xc8f
[2020-02-12T08:16:31.405Z] FAIL
[2020-02-12T08:16:31.405Z] FAIL	github.com/pingcap/tidb/executor	13.710s
```

2. What did you expect to see?

3. What did you see instead?

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

https://github.com/pingcap/tidb/blob/56438a04cd225a7c7924e3cedb486a11159f6e4e/executor/cluster_reader.go#L534

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

This PR will fix #14747.

There will be goroutine  leak in the following situation in `clusterLogRetriever`:

1. Open a log search stream on a remote target and receive `SearchLogResponse` continuously.
2. Executor receives a `SearchLogResponse` from the channel in one retrieval iteration.
3. The retrieval iteration consumes part of lines in the `SearchLogResponse` (receive from channel again just if all lines in the `SearchLogResponse` be consumed).
4. The `MemTableReaderExec`(`clusterLogRetriever`) be closed by parent.
5. The stream receiver goroutine will be leaked.

### What is changed and how it works?

Add a new `exit` channel and check whether is `exit` channel has been closed.

### Check List 

Tests 

 - Unit test

Release note

 - Fix the issue that potential goroutine leak in cluster log retriever

### CODE REVIEW

File: executor/memtable_reader.go [hunk 1]
```diff
@@ -531,10 +536,19 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 						return
 					}
 					if err != nil {
-						ch <- logStreamResult{addr: address, typ: serverType, err: err}
+						select {
+						case ch <- logStreamResult{addr: address, typ: serverType, err: err}:
+						case <-exit:
```
Comment (djshow832) [inline executor/memtable_reader.go]: Why not check it by `ctx.Done()`, but use an extra variable `exit` instead?

Comment (lonng) [reply executor/memtable_reader.go]: Nice catch. addressed.

File: executor/memtable_reader.go [hunk 2]
```diff
@@ -531,10 +539,19 @@ func (e *clusterLogRetriever) startRetrieving(ctx context.Context, sctx sessionc
 						return
 					}
 					if err != nil {
-						ch <- logStreamResult{addr: address, typ: serverType, err: err}
+						select {
+						case ch <- logStreamResult{addr: address, typ: serverType, err: err}:
+						case <-ctx.Done():
```
Comment (djshow832) [inline executor/memtable_reader.go]: Maybe `ctx.Done()` should be front of `ch <- logStreamResult`, to make `ctx.Done()` the first choice when they're both ready?

Comment (lonng) [reply executor/memtable_reader.go]: It's doesn't matter here because the `ch` has no buffer.

**Pull request merged**